pytential/symbolic/primitives.py

```diff
@@ -1123,6 +1152,34 @@ def weights_and_area_elements(ambient_dim, dim=None, dofdesc=None):
 # }}}
 
 
+# {{{ flattened quantities
+
+def _flat_nodes(ambient_dim, dofdesc=None):
+    return cse(
+            _Flatten(
+                nodes(ambient_dim, dofdesc=dofdesc).as_vector(),
+                dofdesc),
+            scope=cse_scope.DISCRETIZATION)
+
+
+def _flat_expansion_radii(ambient_dim, dim=None, dofdesc=None):
+    return cse(
+            _Flatten(
+                expansion_radii(ambient_dim, dim=dim, dofdesc=dofdesc),
+                dofdesc),
+            scope=cse_scope.DISCRETIZATION)
+
+
+def _flat_expansion_centers(ambient_dim, side, dim=None, dofdesc=None):
+    return cse(
+            _Flatten(
+                expansion_centers(ambient_dim, side, dim=dim, dofdesc=dofdesc),
+                dofdesc),
+            scope=cse_scope.DISCRETIZATION)
```

Comment:
I get *why* these exist, but they're annoying IMO. They inflict management of a concern on the user (flattening) that the infrastructure should be able to figure out without paying a massive penalty. If we made use of the fact that (in many common cases) flattening is a reshape, would that allow us to not do this and still dodge most of the flattening cost?

Reply:
I'm really starting to dislike this too, at least the symbolic part. I'll go ahead and remove it for now.

Reply:
Removed in 1a2338d.

It seems to add about 1s (out of a total of 9s-ish) to the case I was playing with. I imagine that'll get worse for larger geometries / more groups. 

How do you feel about just adding a `memoize_in(self)` in `exec_compute_potential_insn_direct` for this? The FMM side seems to do its own flattening in the `GeometryData` classes, so it doesn't feel like it's useful to do something too general here.

Reply:
I'm OK with that.

Reply:
Added in 43a9d9a.

They're memoized in the geometry collection that's passed in.